Added refresh rate for picture entity card

kkellner opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
The hui-image used in picture-entity card has a hard-coded refresh rate of 10 seconds.  To provide more flexibility in the image refresh rate a new user-configurable attribute has been added to the picture-entity card.

## Type of change
<!--
  What type of change does your PR introduce to the Home Assistant frontend?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR.
-->
For `picture-entity` card add a `refresh_rate` property.  This is optional and defaults to the original behavior of 10 seconds if the attribute is not defined.

```yaml
aspect_ratio: 70%
entity: camera.blueiris_cam24
show_name: false
show_state: false
type: picture-entity
refresh_rate: 2
```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- Link to documentation pull request:  https://github.com/home-assistant/home-assistant.io/pull/12710

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

<!--
  Thank you for contributing <3
-->
[docs-repository]: https://github.com/home-assistant/home-assistant.io

kkellner committed:
Added refresh rate for picture entity

homeassistant commented:
Hi @kkellner, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;frontend#5466).

Once you do that we will be able to review and accept this pull request.

Thanks!

kkellner committed:
Merge branch 'dev' into dev

homeassistant commented:
Hi @kkellner, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;frontend#5466).

Once you do that we will be able to review and accept this pull request.

Thanks!

kkellner committed:
Fix lint issue var to const

kkellner committed:
Merge branch 'dev' of github.com:kkellner/frontend into dev

bramkragten commented on src/panels/lovelace/components/hui-image.ts:
```diff
@@ -173,7 +175,8 @@ export class HuiImage extends LitElement {
       this._cameraUpdater = window.setInterval(
         () => this._updateCameraImageSrc(),
-        UPDATE_INTERVAL
+        Number(this.refreshRate ? this.refreshRate : DEFAULT_REFRESH_RATE) *
+          1000
       );
```
Don't do the `Number` conversion, the values are numbers

kkellner replied:
I originally coded it with "number", however the value was never saved.  Maybe I missed somewhere I needed to change.

bramkragten commented on src/panels/lovelace/components/hui-image.ts:
```diff
@@ -40,6 +40,8 @@ export class HuiImage extends LitElement {
 
   @property() public aspectRatio?: string;
 
+  @property() public refreshRate?: string;
```
This should be a number, not a string

kkellner replied:
I originally coded it with "number", however the value was never saved. Maybe I missed somewhere I needed to change.

bramkragten commented on src/panels/lovelace/components/hui-image.ts:
```diff
@@ -173,7 +175,8 @@ export class HuiImage extends LitElement {
       this._cameraUpdater = window.setInterval(
         () => this._updateCameraImageSrc(),
-        UPDATE_INTERVAL
+        Number(this.refreshRate ? this.refreshRate : DEFAULT_REFRESH_RATE) *
+          1000
```
```suggestion
        (this.refreshRate || DEFAULT_REFRESH_RATE) * 1000
```

bramkragten commented on src/panels/lovelace/components/hui-image.ts:
```diff
@@ -208,9 +211,13 @@ export class HuiImage extends LitElement {
     }
 
+    const cacheTime =
+      (Number(this.refreshRate ? this.refreshRate : DEFAULT_REFRESH_RATE) - 1) *
```
Same as above

bramkragten commented on src/panels/lovelace/editor/config-elements/hui-picture-entity-card-editor.ts:
```diff
@@ -36,6 +36,7 @@ const cardConfigStruct = struct({
   camera_image: "string?",
   camera_view: "string?",
   aspect_ratio: "string?",
+  refresh_rate: "string?",
```
number

bramkragten commented on src/panels/lovelace/editor/config-elements/hui-picture-entity-card-editor.ts:
```diff
@@ -79,6 +80,10 @@ export class HuiPictureEntityCardEditor extends LitElement
   }
 
+  get _refresh_rate(): string {
+    return this._config!.refresh_rate || "10";
```
```suggestion
    return this._config!.refresh_rate || 10;
```

bramkragten commented on src/panels/lovelace/editor/config-elements/hui-picture-entity-card-editor.ts:
```diff
@@ -79,6 +80,10 @@ export class HuiPictureEntityCardEditor extends LitElement
     return this._config!.aspect_ratio || "50";
   }
 
+  get _refresh_rate(): string {
```
```suggestion
  get _refresh_rate(): number {
```

bramkragten commented on src/panels/lovelace/editor/config-elements/hui-picture-entity-card-editor.ts:
```diff
@@ -188,6 +193,19 @@ export class HuiPictureEntityCardEditor extends LitElement
+              "ui.panel.lovelace.editor.card.config.optional"
+            )})"
+            type="number"
+            .value="${this._refresh_rate}"
+            .configValue="${"refresh_rate"}"
+            @value-changed="${this._valueChanged}"
```
```suggestion
            .value=${this._refresh_rate}
            .configValue=${"refresh_rate"}
            @value-changed=${this._valueChanged}
```

bramkragten commented on src/panels/lovelace/elements/hui-image-element.ts:
```diff
@@ -52,6 +52,7 @@ export class HuiImageElement extends LitElement implements LovelaceElement {
         .stateFilter="${this._config.state_filter}"
         .title="${computeTooltip(this.hass, this._config)}"
         .aspectRatio="${this._config.aspect_ratio}"
+        .refreshRate="${this._config.refresh_rate}"
```
```suggestion
        .refreshRate=${this._config.refresh_rate}
```

bramkragten commented on src/panels/lovelace/elements/types.ts:
```diff
@@ -46,6 +46,7 @@ export interface ImageElementConfig extends LovelaceElementConfigBase {
   filter?: string;
   state_filter?: string;
   aspect_ratio?: string;
+  refresh_rate?: string;
```
```suggestion
  refresh_rate?: number;
```

bramkragten commented on src/translations/en.json:
```diff
@@ -2070,6 +2070,7 @@
             },
             "generic": {
               "aspect_ratio": "Aspect Ratio",
+              "refresh_rate": "Refresh Rate (seconds)",
```
Translations are handled in Localize, you should not add them here.

bramkragten commented on translations/ca.json:
```diff
@@ -2178,6 +2178,7 @@
                         },
                         "generic": {
                             "aspect_ratio": "Relació d'aspecte",
+                            "refresh_rate": "Velocitat d’actualització (segons)",
```
Translations are handled in Localize, you should not add them here.

kkellner replied:
This is my first time contributing to HA -- I didn't find any other finds that contained the translations.   When you say "handled in Localize", what do you mean?  Is there a different git repo for this?

zsarnett replied:
src\translations\en.json

This is the only place that it needs to be added

kkellner replied:
Ok, I get it now.  Thanks for pointing that out.

zsarnett replied:
Please remove the other translation

kkellner replied:
Removed

bramkragten commented:
I don't think this is a good idea, a user can flood his Home Assistant now. We have `camera_view: live` for this.

kkellner commented:
The `live` view on my setup has a 7-12 second delay between showing on browser and what is "really live".   I am using Blue Iris -- when I go directly to the web interface of Blue Iris its immediate.  Not sure if there is buffering occuring in HA.   However, the image view (auto) was instant.   I agree that using low image refresh rates (e.g., 1 second) should be used with caution.

kkellner committed:
Update src/panels/lovelace/components/hui-image.ts

Co-Authored-By: Bram Kragten <mail@bramkragten.nl>

kkellner committed:
Update src/panels/lovelace/editor/config-elements/hui-picture-entity-card-editor.ts

Co-Authored-By: Bram Kragten <mail@bramkragten.nl>

kkellner committed:
Update src/panels/lovelace/editor/config-elements/hui-picture-entity-card-editor.ts

Co-Authored-By: Bram Kragten <mail@bramkragten.nl>

kkellner committed:
Update src/panels/lovelace/elements/types.ts

Co-Authored-By: Bram Kragten <mail@bramkragten.nl>

kkellner committed:
Update src/panels/lovelace/editor/config-elements/hui-picture-entity-card-editor.ts

Co-Authored-By: Bram Kragten <mail@bramkragten.nl>

kkellner committed:
Update src/panels/lovelace/elements/hui-image-element.ts

Co-Authored-By: Bram Kragten <mail@bramkragten.nl>

kkellner committed:
Change refreshRate to number

kkellner committed:
Set to NaN vs 10

kkellner commented:
Changed everything to number and it is working.  The only thing that wasn't working well is when, for example, the user changed the default of 10 to 5, the user would press the delete key twice to delete "10" before entering "5".   When the field is empty it resulted in the field getting initialized back to 10.  I changed the getter to return NaN instead of 10 when value is not specified by user which then uses the internal default of 10.   Since this is an optional field, this seemed acceptable.

[NaN Change](https://github.com/home-assistant/frontend/pull/5466/commits/e46d0779a8cdaf8fcefbb79d6d4fba8597f496a4)

kkellner committed:
Remove refresh_rate translations

zsarnett commented on translations/ca.json:
```diff
@@ -2164,6 +2164,7 @@
                             "name": "Filtre d'entitats"
                         },
                         "entity": {
+                            "description": "La targeta entitat mostra una visualització ràpida dels estats d'una o més entitats.",
```
All translations will need removed except for `src/translations/en.json`

kkellner replied:
I believe I fixed.

zsarnett replied:
Good now

zsarnett commented:
Maybe we can limit this to being 10 seconds or greater? If the user makes it lower than 10 it is changed to being 10

kkellner commented:
@zsarnett The original use case was to set the update interval to lower then 10 seconds as the live streaming from Blue Iris that flows through HA has a 7+ second delay.   So the work around was to refresh a static image every 3 seconds.   Maybe we can do some type of warning if the value is less than 10 seconds to inform user that the performance may be impacted.

zsarnett commented:
Are you able to test and see if that work? Refreshing at a 3 second delay? IS HA now more in sync?

I know I see a delay with RTSP but I am not sure the delay is HA or just the RTSP

kkellner commented:
I can have a 1 second refresh rate on a modern laptop and it keeps up, but the target is a Fire Tablet which is slow, so backed off to 3 seconds which seems to work well.

As for live streaming delay.   I can see that there is a 3 second delay coming from Blue Iris on the type of stream that is requested (m3u8 MIME type application/vnd.apple.mpegurl).  I think its because it has to transcode to get this format.   The additional delay seems to happen between between HA getting the stream and the browser getting the steaming bytes.  This is something I want to dig in more to see if the native "video/mpeg" content type can be used from Blue Iris as this is much faster -- but that's a different issue.

bramkragten commented:
@kkellner I still don't think this is a good idea, if we would add this, it should not be in the UI editor as it is too advanced and could break systems.

Would it be an idea to create a custom card for this?

kkellner commented:
@bramkragten It doesn't seem like a refresh rate is too advanced of an option.  Are you thinking that users will change to a value that is too low that HA would not be able to keep up?    I was trying to avoid creating another card type in order to keep the functionality together and the code in sync with what is already provided.  If you think this isn't going to workout, you can toss the change.

nagyrobi commented:
3 seconds is perfect here. 10 seconds too slow.
I was looking for a way to speed up.
